File: services/graphql-dms/src/mutation/driver_locations.rs
use async_graphql::{Context, InputObject, Object};
use uuid::Uuid;

#[derive(Debug, Clone, InputObject)]
pub struct CreateDriverLocationInput {
    pub driver_id: Uuid,
    pub latitude: f32,
    pub longitude: f32,
    pub altitude: Option<f32>,
    pub accuracy: Option<f32>,
    pub speed_kmh: Option<f32>,
    pub heading: Option<f32>,
    pub timestamp: Option<chrono::DateTime<chrono::Utc>>,
}

#[derive(Debug, Clone, Default)]
pub struct Mutation;

#[Object(name = "DmsDriverLocationsMutations")]
impl Mutation {
    async fn create_driver_location(
        &self,
        ctx: &Context<'_>,
        payload: CreateDriverLocationInput,
    ) -> async_graphql::Result<crate::models::driver_locations::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::driver_locations::Model>(
            "insert into dms.driver_locations (driver_id, latitude, longitude, altitude, accuracy, speed_kmh, heading, timestamp) values ($1, $2, $3, $4, $5, $6, $7, $8) returning *"
        )
        .bind(payload.driver_id)
        .bind(payload.latitude)
        .bind(payload.longitude)
        .bind(payload.altitude)
        .bind(payload.accuracy)
        .bind(payload.speed_kmh)
        .bind(payload.heading)
        .bind(payload.timestamp)
        .fetch_one(db)
        .await?)
    }

    async fn update_driver_location_driver_id(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        driver_id: Uuid,
    ) -> async_graphql::Result<crate::models::driver_locations::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::driver_locations::Model>(
            "update dms.driver_locations set driver_id = $1 where id = $2 returning *",
        )
        .bind(driver_id)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_driver_location_latitude(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        latitude: f32,
    ) -> async_graphql::Result<crate::models::driver_locations::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::driver_locations::Model>(
            "update dms.driver_locations set latitude = $1 where id = $2 returning *",
        )
        .bind(latitude)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_driver_location_longitude(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        longitude: f32,
    ) -> async_graphql::Result<crate::models::driver_locations::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::driver_locations::Model>(
            "update dms.driver_locations set longitude = $1 where id = $2 returning *",
        )
        .bind(longitude)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_driver_location_altitude(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        altitude: Option<f32>,
    ) -> async_graphql::Result<crate::models::driver_locations::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::driver_locations::Model>(
            "update dms.driver_locations set altitude = $1 where id = $2 returning *",
        )
        .bind(altitude)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_driver_location_accuracy(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        accuracy: Option<f32>,
    ) -> async_graphql::Result<crate::models::driver_locations::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::driver_locations::Model>(
            "update dms.driver_locations set accuracy = $1 where id = $2 returning *",
        )
        .bind(accuracy)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_driver_location_speed_kmh(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        speed_kmh: Option<f32>,
    ) -> async_graphql::Result<crate::models::driver_locations::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::driver_locations::Model>(
            "update dms.driver_locations set speed_kmh = $1 where id = $2 returning *",
        )
        .bind(speed_kmh)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_driver_location_heading(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        heading: Option<f32>,
    ) -> async_graphql::Result<crate::models::driver_locations::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::driver_locations::Model>(
            "update dms.driver_locations set heading = $1 where id = $2 returning *",
        )
        .bind(heading)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_driver_location_timestamp(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        timestamp: Option<chrono::DateTime<chrono::Utc>>,
    ) -> async_graphql::Result<crate::models::driver_locations::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::driver_locations::Model>(
            "update dms.driver_locations set timestamp = $1 where id = $2 returning *",
        )
        .bind(timestamp)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_driver_location_position(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        latitude: f32,
        longitude: f32,
        altitude: Option<f32>,
    ) -> async_graphql::Result<crate::models::driver_locations::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::driver_locations::Model>(
            "update dms.driver_locations set latitude = $1, longitude = $2, altitude = $3 where id = $4 returning *",
        )
        .bind(latitude)
        .bind(longitude)
        .bind(altitude)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn remove_driver_location(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
    ) -> async_graphql::Result<String> {
        let db = ctx.data::<sqlx::PgPool>()?;
        let result = sqlx::query("delete from dms.driver_locations where id = $1")
            .bind(id)
            .execute(db)
            .await?;
        if result.rows_affected() != 1 {
            return Err(async_graphql::Error::new(
                "Unable to delete driver location",
            ));
        }
        Ok("Driver location removed successfully".into())
    }
}

File: src/graphql/mutations/dms/driverLocations.ts
import { graphql } from "@/lib/graphql/client";

// ============================================================================
// DRIVER LOCATION MUTATIONS
// ============================================================================

/**
 * Creates a new driver location record.
 * @param payload The input data for creating the driver location.
 */
export const createDriverLocation = graphql(`
  mutation CreateDriverLocation($payload: CreateDriverLocationInput!) {
    dms {
      createDriverLocation(payload: $payload) {
        id
        latitude
        longitude
        altitude
        accuracy
        speedKmh
        heading
        timestamp
        createdAt
        updatedAt
        driver {
          id
          licenseNumber
        }
      }
    }
  }
`);

/**
 * Updates the driver ID of a driver location record.
 * @param id The UUID of the driver location record to update.
 * @param driverId The new driver ID.
 */
export const updateDriverLocationDriverId = graphql(`
  mutation UpdateDriverLocationDriverId($id: UUID!, $driverId: UUID!) {
    dms {
      updateDriverLocationDriverId(id: $id, driverId: $driverId) {
        id
        latitude
        longitude
        altitude
        accuracy
        speedKmh
        heading
        timestamp
        createdAt
        updatedAt
        driver {
          id
          licenseNumber
        }
      }
    }
  }
`);

/**
 * Updates the latitude of a driver location record.
 * @param id The UUID of the driver location record to update.
 * @param latitude The new latitude.
 */
export const updateDriverLocationLatitude = graphql(`
  mutation UpdateDriverLocationLatitude($id: UUID!, $latitude: Float!) {
    dms {
      updateDriverLocationLatitude(id: $id, latitude: $latitude) {
        id
        latitude
        longitude
        altitude
        accuracy
        speedKmh
        heading
        timestamp
        createdAt
        updatedAt
        driver {
          id
          licenseNumber
        }
      }
    }
  }
`);

/**
 * Updates the longitude of a driver location record.
 * @param id The UUID of the driver location record to update.
 * @param longitude The new longitude.
 */
export const updateDriverLocationLongitude = graphql(`
  mutation UpdateDriverLocationLongitude($id: UUID!, $longitude: Float!) {
    dms {
      updateDriverLocationLongitude(id: $id, longitude: $longitude) {
        id
        latitude
        longitude
        altitude
        accuracy
        speedKmh
        heading
        timestamp
        createdAt
        updatedAt
        driver {
          id
          licenseNumber
        }
      }
    }
  }
`);

/**
 * Updates the altitude of a driver location record.
 * @param id The UUID of the driver location record to update.
 * @param altitude The new altitude.
 */
export const updateDriverLocationAltitude = graphql(`
  mutation UpdateDriverLocationAltitude($id: UUID!, $altitude: Float) {
    dms {
      updateDriverLocationAltitude(id: $id, altitude: $altitude) {
        id
        latitude
        longitude
        altitude
        accuracy
        speedKmh
        heading
        timestamp
        createdAt
        updatedAt
        driver {
          id
          licenseNumber
        }
      }
    }
  }
`);

/**
 * Updates the accuracy of a driver location record.
 * @param id The UUID of the driver location record to update.
 * @param accuracy The new accuracy.
 */
export const updateDriverLocationAccuracy = graphql(`
  mutation UpdateDriverLocationAccuracy($id: UUID!, $accuracy: Float) {
    dms {
      updateDriverLocationAccuracy(id: $id, accuracy: $accuracy) {
        id
        latitude
        longitude
        altitude
        accuracy
        speedKmh
        heading
        timestamp
        createdAt
        updatedAt
        driver {
          id
          licenseNumber
        }
      }
    }
  }
`);

/**
 * Updates the speed in km/h of a driver location record.
 * @param id The UUID of the driver location record to update.
 * @param speedKmh The new speed in km/h.
 */
export const updateDriverLocationSpeedKmh = graphql(`
  mutation UpdateDriverLocationSpeedKmh($id: UUID!, $speedKmh: Float) {
    dms {
      updateDriverLocationSpeedKmh(id: $id, speedKmh: $speedKmh) {
        id
        latitude
        longitude
        altitude
        accuracy
        speedKmh
        heading
        timestamp
        createdAt
        updatedAt
        driver {
          id
          licenseNumber
        }
      }
    }
  }
`);

/**
 * Updates the heading of a driver location record.
 * @param id The UUID of the driver location record to update.
 * @param heading The new heading.
 */
export const updateDriverLocationHeading = graphql(`
  mutation UpdateDriverLocationHeading($id: UUID!, $heading: Float) {
    dms {
      updateDriverLocationHeading(id: $id, heading: $heading) {
        id
        latitude
        longitude
        altitude
        accuracy
        speedKmh
        heading
        timestamp
        createdAt
        updatedAt
        driver {
          id
          licenseNumber
        }
      }
    }
  }
`);

/**
 * Updates the timestamp of a driver location record.
 * @param id The UUID of the driver location record to update.
 * @param timestamp The new timestamp.
 */
export const updateDriverLocationTimestamp = graphql(`
  mutation UpdateDriverLocationTimestamp($id: UUID!, $timestamp: DateTime) {
    dms {
      updateDriverLocationTimestamp(id: $id, timestamp: $timestamp) {
        id
        latitude
        longitude
        altitude
        accuracy
        speedKmh
        heading
        timestamp
        createdAt
        updatedAt
        driver {
          id
          licenseNumber
        }
      }
    }
  }
`);

/**
 * Updates the position (latitude, longitude, and altitude) of a driver location record.
 * @param id The UUID of the driver location record to update.
 * @param latitude The new latitude.
 * @param longitude The new longitude.
 * @param altitude The new altitude.
 */
export const updateDriverLocationPosition = graphql(`
  mutation UpdateDriverLocationPosition($id: UUID!, $latitude: Float!, $longitude: Float!, $altitude: Float) {
    dms {
      updateDriverLocationPosition(id: $id, latitude: $latitude, longitude: $longitude, altitude: $altitude) {
        id
        latitude
        longitude
        altitude
        accuracy
        speedKmh
        heading
        timestamp
        createdAt
        updatedAt
        driver {
          id
          licenseNumber
        }
      }
    }
  }
`);

/**
 * Removes a driver location record by its ID.
 * @param id The UUID of the driver location record to remove.
 */
export const removeDriverLocation = graphql(`
  mutation RemoveDriverLocation($id: UUID!) {
    dms {
      removeDriverLocation(id: $id)
    }
  }
`);
